### PR TITLE
feat(libflux/core/parser): update mod.rs with parse_constraints

### PR DESCRIPTION
**Pull request by yuktarakesh**
add struct Constraints, edit struct Type_Expression with constraints and add corresponding unit test.

### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

### CODE REVIEW

File: libflux/src/core/parser/mod.rs [hunk 1]
```diff
@@ -590,6 +591,38 @@ impl Parser {
...
+        while self.peek().tok == TOK_COMMA {
+            self.consume();
+            constraints.push(self.parse_constraint());
+        }
+        return constraints;
+    }
+
+    #[cfg(test)]
+    fn parse_constraint(&mut self) -> Constraints {
+        let t = self.peek();
```
Comment (jlapacik) [inline libflux/src/core/parser/mod.rs:607]: I don't think you need this line here.

File: libflux/src/core/parser/mod.rs [hunk 2]
```diff
@@ -590,6 +591,38 @@ impl Parser {
...
+        self.expect(TOK_COLON);
+        let identifier = self.parse_identifier();
+        id.push(identifier);
+        while self.peek().tok == TOK_ADD {
+            self.consume();
+            let identifier = self.parse_identifier();
+            id.push(identifier);
+        }
+        let con = Constraints {
+            base: self.base_node_from_token(&t),
```
Comment (jlapacik) [inline libflux/src/core/parser/mod.rs:619]: I think this should be `self.base_node_from_tokens(_tvar, end)` where `end` is the token representing the last identifier.

File: libflux/src/core/parser/tests.rs [hunk 3]
```diff
@@ -1531,11 +1542,42 @@ fn test_parse_type_expression_array_string() {
...
+#[test]
+fn test_parse_constraints() {
+    let mut p = Parser::new(r#"A : date"#);
+    let parsed = p.parse_constraints();
+    let loc = Locator::new(&p.source[..]);
+    assert_eq!(
+        parsed,
+        vec![Constraints {
+            base: BaseNode {
+                location: loc.get(1, 1, 1, 2),
```
Comment (jlapacik) [inline libflux/src/core/parser/tests.rs:1560]: Should be `loc.get(1,1,1,9)`

File: libflux/src/core/ast/mod.rs [hunk 4]
```diff
@@ -542,8 +542,18 @@ pub struct TypeExpression {
     #[serde(flatten)]
     pub base: BaseNode,
     pub monotype: MonoType,
+    pub constraints: Option<Constraints>,
 }
 
+#[derive(Debug, PartialEq, Clone, Serialize, Deserialize)]
+pub struct Constraints {
```
Comment (nathanielc) [inline libflux/src/core/ast/mod.rs:549]: Should we call this `TypeConstraints`?

Comment (yuktarakesh) [reply libflux/src/core/ast/mod.rs]: I think we can make it ConstraintType to match the others?

**Pull request merged**